enginetest/enginetests.go

```diff
@@ -3174,62 +3173,129 @@ func TestWindowRowFrames(t *testing.T, harness Harness) {
 		`SELECT sum(y) over (partition by z order by x rows between 2 preceding and 1 preceding) FROM a order by x`,
 		[]sql.Row{{nil}, {float64(0)}, {float64(1)}, {float64(3)}, {float64(2)}, {float64(1)}},
 		nil, nil)
+}
 
-	// TODO implement ranges
-	// ranges
-	//RunQuery(t, e, harness, "CREATE TABLE b (x INTEGER PRIMARY KEY, y INTEGER, z INTEGER)")
-	//RunQuery(t, e, harness, "INSERT INTO a VALUES (0,0,0), (1,0,0), (2,0,0), (3,1,0), (4,1,0), (5,3,0)")
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range 2 following) FROM a order by x`,
-	//	[]sql.Row{{0},{0},{0},{0},{0},{0}},
-	//	nil, nil)
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range unbounded preceding) FROM a order by x`,
-	//	[]sql.Row{{0},{1},{3},{3},{4},{7}},
-	//	nil, nil)
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range current row) FROM a order by x`,
-	//	[]sql.Row{{0},{1},{2},{0},{1},{3}},
-	//	nil, nil)
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range 2 preceding) FROM a order by x`,
-	//	[]sql.Row{{0},{1},{3},{3},{3},{4}},
-	//	nil, nil)
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range between current row and 1 following) FROM a order by x`,
-	//	[]sql.Row{{1},{3},{2},{1},{4},{3}},
-	//	nil, nil)
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range between 1 preceding and current row) FROM a order by x`,
-	//	[]sql.Row{{0},{1},{3},{2},{1},{4}},
-	//	nil, nil)
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range between current row and 2 following) FROM a order by x`,
-	//	[]sql.Row{{3},{3},{3},{4},{4},{3}},
-	//	nil, nil)
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range between current row and current row) FROM a order by x`,
-	//	[]sql.Row{{0},{1},{2},{0},{1},{3}},
-	//	nil, nil)
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range between current row and unbounded following) FROM a order by x`,
-	//	[]sql.Row{{7},{7},{6},{4},{4},{3}},
-	//	nil, nil)
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range between 1 preceding and 1 following) FROM a order by x`,
-	//	[]sql.Row{{1},{3},{3},{3},{4},{4}},
-	//	nil, nil)
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range between 1 preceding and unbounded following) FROM a order by x`,
-	//	[]sql.Row{{7},{7},{7},{6},{4},{4}},
-	//	nil, nil)
-	//TestQuery(t, harness, e,
-	//	`SELECT sum(y) over (partition by z order by x range between unbounded preceding and unbounded following) FROM a order by x`,
-	//	[]sql.Row{{7},{7},{7},{7},{7},{7}},
-	//	nil, nil)
+func TestWindowRangeFrames(t *testing.T, harness Harness) {
```

Comment:
Tests of invalid ranges?

You need an order by and it has to be on a numeric expression (or a date expression for interval ranges)